vncviewer/vncviewer.cxx

```diff
@@ -109,28 +108,47 @@ static const char *about_text()
   return buffer;
 }
 
-void exit_vncviewer(const char *error, ...)
+
+void abort_vncviewer(const char *error, ...)
 {
   // Prioritise the first error we get as that is probably the most
   // relevant one.
-  if ((error != NULL) && (exitError == NULL)) {
+  char* errorString = NULL;
+  if (error != NULL) {
     va_list ap;
 
     va_start(ap, error);
-    exitError = (char*)malloc(1024);
-    if (exitError)
-      (void) vsnprintf((char*)exitError, 1024, error, ap);
+    errorString = (char*)malloc(1024);
+    if (errorString) {
+      (void) vsnprintf((char*)errorString, 1024, error, ap);
+      vlog.error("%s",errorString);
+    }
     va_end(ap);
   }
 
-  if (inMainloop)
-    exitMainloop = true;
```

Comment:
This function could unfortunately be called when the mainloop has been started, so this still needs to be handled.

Reply:
That's my point, it shouldn't be calling `exit()` and just abruptly terminating things. At least not if we've gotten to the point of the mainloop. Getting the ability to clean things up is always preferred.

Reply:
`exit_vncviewer` is called before the main loop is entered in vncviewer.cxx. Only in touch.cxx it might be called when we are already in the main loop and I read the code like it should have all cases a terminating character (`return; //Not reached` is directly after the call to `exit_vncviewer`). So I see no need for a graceful shutdown. While writing this maybe the better name is to call this `terminate_vncviewer` to make it clear that this is not a graceful shutdown.

Reply:
That comment was probably just a copy-paste error.

Shutting down as gracefully as we can should always be preferred unless there is some technical limitations preventing us from doing so. Otherwise we might leave stray stuff around, not signal other programs properly, or various other issues.

Reply:
Than my proposal would be to limit `exit_vncviewer` to vncviewer.cxx and have all other places to call `abort_vncviewer` which does a graceful termination/reconnect. Is this the direction you are thinking?

Reply:
No, I still think it's a good separation of one aborting the connection, and the other aborting the viewer. There are some errors where a reconnect will not be able to fix things (hopefully few though).

So this function here should be able to break the mainloop in a way that does not trigger a reconnect. A simple extra flag should be able to solve that.